assertion fail with TwoBit reference (query past end of contig)

command 
```
spark-submit \
  --master $SPARK_MASTER \
  --driver-memory 8G \
  --num-executors ${1} \
  --executor-cores 4 \
  --executor-memory 20G \
  --conf spark.driver.maxResultSize=0 \
  --conf spark.driver.userClassPathFirst=true \
  --conf spark.executor.userClassPathFirst=true \
  --conf spark.io.compression.codec=lzf \
  --conf spark.yarn.executor.memoryOverhead=600 \
      $JAR BaseRecalibratorSpark \
        --input hdfs:///user/akiezun/CEUTrio.HiSeq.WEx.b37.NA12892.bam \
        --output bqsr_out_${1}.bam \
        -R hdfs:///user/droazen/bqsr/human_g1k_v37.2bit \
        --knownSites hdfs:////user/akiezun/dbsnp_138.b37.excluding_sites_after_129.vcf \
        --joinStrategy BROADCAST \
        --apiKey $API_KEY \
        --sparkMaster $SPARK_MASTER
```

fails with
```
java.lang.AssertionError: assertion failed
	at scala.Predef$.assert(Predef.scala:165)
	at org.bdgenomics.adam.util.TwoBitFile.extract(TwoBitFile.scala:112)
	at org.bdgenomics.adam.util.TwoBitFile.extract(TwoBitFile.scala:170)
	at org.broadinstitute.hellbender.engine.spark.datasources.ReferenceTwoBitSource.getReferenceBases(ReferenceTwoBitSource.java:48)
	at org.broadinstitute.hellbender.engine.datasources.ReferenceMultiSource.getReferenceBases(ReferenceMultiSource.java:99)
	at org.broadinstitute.hellbender.engine.spark.BroadcastJoinReadsWithRefBases.lambda$addBases$c54addeb$1(BroadcastJoinReadsWithRefBases.java:53)
...
```

--- a/src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReferenceTwoBitSource.java
+++ b/src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReferenceTwoBitSource.java
@@ -30,8 +30,9 @@ public class ReferenceTwoBitSource implements ReferenceSource, Serializable {
 
     public static final String TWO_BIT_EXTENSION = ".2bit";
 
-    private String referenceURL;
-    private TwoBitFile twoBitFile;
+    private final String referenceURL;
+    private final TwoBitFile twoBitFile;
+    private final Map<String, TwoBitRecord> twoBitSeqEntries;
 
     public ReferenceTwoBitSource(PipelineOptions popts, String referenceURL) throws IOException {
         this.referenceURL = referenceURL;
@@ -41,18 +42,28 @@ public class ReferenceTwoBitSource implements ReferenceSource, Serializable {
         byte[] bytes = ByteStreams.toByteArray(BucketUtils.openFile(this.referenceURL, popts));
         ByteAccess byteAccess = new DirectFullByteArrayByteAccess(bytes);
         this.twoBitFile = new TwoBitFile(byteAccess);
+        this.twoBitSeqEntries = JavaConversions.mapAsJavaMap(twoBitFile.seqRecords());
     }
 
+    /**
+     * Gets the reference bases spanning the requested interval. If the interval ends beyond the end of its
+     * contig according to our reference source's dictionary, it will be truncated at the contig end.
+     *
+     * @param pipelineOptions pipeline options (may be null)
+     * @param interval query interval
+     * @return A ReferenceBases containing the reference bases spanning the requested interval, cropped at the
+     *         contig end if necessary
+     */
     @Override
     public ReferenceBases getReferenceBases(PipelineOptions pipelineOptions, SimpleInterval interval) throws IOException {
-        String bases = twoBitFile.extract(simpleIntervalToReferenceRegion(interval));
-        return new ReferenceBases(bases.getBytes(), interval);
+        final SimpleInterval queryInterval = cropIntervalAtContigEnd(interval);
+        final String bases = twoBitFile.extract(simpleIntervalToReferenceRegion(queryInterval));
+        return new ReferenceBases(bases.getBytes(), queryInterval);
     }
 
     @Override
     public SAMSequenceDictionary getReferenceSequenceDictionary(SAMSequenceDictionary optReadSequenceDictionaryToMatch) throws IOException {
-        Map<String, TwoBitRecord> twoBitEntries = JavaConversions.mapAsJavaMap(twoBitFile.seqRecords());
-        List<SAMSequenceRecord> records = twoBitEntries.entrySet().stream()
+        List<SAMSequenceRecord> records = twoBitSeqEntries.entrySet().stream()
                 .map(pair -> new SAMSequenceRecord(pair.getKey(), pair.getValue().dnaSize()))
                 .collect(Collectors.toList());
         return new SAMSequenceDictionary(records);
@@ -69,4 +80,16 @@ public class ReferenceTwoBitSource implements ReferenceSource, Serializable {
         long end = interval.getGA4GHEnd();
         return new ReferenceRegion(contig, start, end, null);
     }
+
+    private SimpleInterval cropIntervalAtContigEnd( final SimpleInterval interval ) {
+        // The 2bit query API does not support queries beyond the ends of contigs, so we need
+        // to truncate our interval at the contig end if necessary.
+        final TwoBitRecord contigRecord = twoBitSeqEntries.get(interval.getContig());
+        if ( contigRecord == null ) {
+            throw new IllegalArgumentException("Contig " + interval.getContig() + " not found in reference dictionary");
+        }
+
+        return new SimpleInterval(interval.getContig(), interval.getStart(), Math.min(interval.getEnd(), contigRecord.dnaSize()));
+    }
+
 }

--- a/src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReferenceTwoBitSourceUnitTest.java
+++ b/src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReferenceTwoBitSourceUnitTest.java
@@ -34,4 +34,28 @@ public class ReferenceTwoBitSourceUnitTest extends BaseTest {
         ReferenceBases actual = twoBitRef.getReferenceBases(null, interval);
         Assert.assertEquals(actual, expected);
     }
+
+    @DataProvider(name = "outOfBoundsIntervals")
+    public Object[][] getOutOfBoundsIntervals() throws IOException {
+        final ReferenceTwoBitSource twoBitRef = new ReferenceTwoBitSource(null, publicTestDir + "large/human_g1k_v37.20.21.2bit");
+        final int chr20End = 63025520;
+
+        return new Object[][] {
+                { twoBitRef, new SimpleInterval("20", chr20End - 100, chr20End + 1), 101, chr20End },
+                { twoBitRef, new SimpleInterval("20", chr20End - 100, chr20End + 100), 101, chr20End },
+                { twoBitRef, new SimpleInterval("20", chr20End - 1, chr20End + 1), 2, chr20End },
+                { twoBitRef, new SimpleInterval("20", chr20End, chr20End + 1), 1, chr20End },
+        };
+    }
+
+    @Test(dataProvider = "outOfBoundsIntervals")
+    public void testQueryPastContigEnd( final ReferenceTwoBitSource refSource, final SimpleInterval outOfBoundsInterval, final int expectedNumBases, final int contigEnd ) throws IOException {
+        final ReferenceBases bases = refSource.getReferenceBases(null, outOfBoundsInterval);
+
+        // Verify that the ReferenceTwoBitSource cropped our out-of-bounds interval at the contig end, as expected,
+        // and that we got the correct number of bases back.
+        Assert.assertEquals(bases.getInterval().getEnd(), contigEnd, "Interval was not cropped at contig end");
+        Assert.assertEquals(bases.getBases().length, expectedNumBases, "Wrong number of bases returned from query");
+        Assert.assertEquals(bases.getInterval().size(), expectedNumBases, "Wrong interval in ReferenceBases object returned from query");
+    }
 }

